dbms/src/Storages/System/StorageSystemTables.cpp

```diff
@@ -176,16 +191,23 @@ class TablesBlockInputStream : public IBlockInputStream
             if (!tables_it || !tables_it->isValid())
                 tables_it = database->getIterator(context);
 
+            const bool need_lock_structure = needLockStructure(database, header);
+
             for (; rows_count < max_block_size && tables_it->isValid(); tables_it->next())
             {
                 auto table_name = tables_it->name();
-                const StoragePtr & table = tables_it->table();
+                StoragePtr table = nullptr;
 
                 TableStructureReadLockHolder lock;
 
                 try
                 {
-                    lock = table->lockStructureForShare(false, context.getCurrentQueryId());
+                    if (need_lock_structure)
+                    {
+                        if (!table)
+                            table = tables_it->table();
+                        lock = table->lockStructureForShare(false, context.getCurrentQueryId());
```

Comment:
table is always nullptr here. No need in this code, or there is a bug?

Reply:
> table is always nullptr here

Don't understand, could you please elaborate?